### PR TITLE
Add support for fetching rules for multiple tenants

### LINKED ISSUE #7: Add support for multiple tenants
### Problem

Thanos Rule Syncer currently supports fetching rules for a single tenant from Observatorium API. This means that if you are running Obs API with multiple tenants (almost always), you'd need to run one Thanos Rule Syncer sidecar for each tenant.

### Proposed Solution

We can add support for specifying multiple tenants and their credentials as input to Rule Syncer (maybe as a yaml file) and then we can fetch rules for all of them, combine in a single file, and then follow the same pattern we followed for a single tenant.

### PR DESCRIPTION
**Pull request by onprem**
Fixes #7 

- [x] Add basic support for fetching rules for multiple tenants.
- [x] Do networks requests to fetch rules concurrently.
- [ ] Handle tenants with invalid configuration gracefully.
- [ ] Instrument skipped/failed tenants with metrics.

Signed-off-by: Prem Saraswat <prmsrswt@gmail.com>

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -161,6 +186,53 @@ func main() {
...
+
+	go func() {
+		var wg sync.WaitGroup
+		wg.Add(len(tenants))
+
+		for i := 0; i < fetchConcurrency; i++ {
+			go func() {
+				for tenant := range queue {
+					observatoriumURL.Path = path.Join("/api/metrics/v1", tenant.Name, "/api/v1/rules/raw")
+					// TODO(onprem): Handle errors.
```
Comment (squat) [inline main.go:201]: This seems actually pretty crucial. Getting the rules is the core functionality of this whole project; if this fails, we need to know and retry. I think we should address this :+1:

**Review by bill3tt — CHANGES_REQUESTED** [review]
Sorry to block @onprem but I feel quite strongly that we should not be handling, storing or using any user's secret material in this component.

As a sidecar in Ruler, `thanos-ruler-syncer` is already part of a component that communicates directly with the thanos backends and bypasses API auth. It is acceptable for `thanos-ruler-syncer` to do the same.

If `rules-objstore` exposed a `ListAllRules` endpoint, `thanos-ruler-syncer` could call that to get the rules. As the API enforces tenant headers, we could return all rules when the tenant header is not set? 

That is less than ideal, but is a preferable problem to have than managing user secret material.

**Comment by onprem** [comment]
> If rules-objstore exposed a ListAllRules endpoint, thanos-ruler-syncer could call that to get the rules. As the API enforces tenant headers, we could return all rules when the tenant header is not set?

The rules backend utilises URL path parameters instead of tenant header (i.e `/api/v1/<tenant-name>/rules` ) but we can just create another endpoint that lists all rules.

I agree with your point that Rule Syncer is internal and it would be better to skip Obs API auth.

**Comment by onprem** [comment]
Closing in favor of #9

**Pull request closed without merging**